homeassistant/components/awair/config_flow.py

```diff
@@ -0,0 +1,104 @@
+"""Config flow for Awair."""
+
+from typing import Union
+
+from python_awair import Awair
+from python_awair.exceptions import AuthError, AwairError
+import voluptuous as vol
+
+from homeassistant.config_entries import CONN_CLASS_CLOUD_POLL, ConfigFlow
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN, LOGGER  # pylint: disable=unused-import
+
+
+class AwairFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Config flow for Awair."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_CLOUD_POLL
+
+    async def async_step_import(self, conf: dict):
+        """Import a configuration from config.yaml."""
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason="already_setup")
+
+        if not conf[CONF_ACCESS_TOKEN]:
+            return self.async_abort(reason="auth")
+
+        await self._abort_if_configured(conf[CONF_ACCESS_TOKEN])
+
+        user, errors = await self._check_access_token(conf[CONF_ACCESS_TOKEN])
+        if errors is not None:
+            return self.async_abort(reason="auth")
+
+        return self.async_create_entry(
+            title=f"{user.email} ({user.user_id})",
+            data={CONF_ACCESS_TOKEN: conf[CONF_ACCESS_TOKEN]},
+        )
+
+    async def async_step_user(self, user_input: Union[dict, None] = None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if user_input is not None:
+            user, errors = await self._check_access_token(user_input[CONF_ACCESS_TOKEN])
+
+            if not errors:
+                await self._abort_if_configured(user_input[CONF_ACCESS_TOKEN])
+                title = f"{user.email} ({user.user_id})"
+                return self.async_create_entry(title=title, data=user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema({vol.Required(CONF_ACCESS_TOKEN): str}),
+            errors=errors,
+        )
+
+    async def async_step_reauth(self, user_input: Union[dict, None] = None):
+        """Handle re-auth if token invalid."""
+        errors = {}
+
+        if user_input is not None:
+            if "config_entry" in user_input:
+                if self.unique_id is None:
+                    await self.async_set_unique_id(user_input["config_entry"].unique_id)
+            elif CONF_ACCESS_TOKEN in user_input:
+                access_token = user_input[CONF_ACCESS_TOKEN]
+                user, errors = await self._check_access_token(access_token)
+                if not errors:
+                    return self.async_create_entry(
+                        title=f"{user.email} ({user.user_id})",
```

Comment:
Why should we create a new entry? What has changed besides the token? If we just want to update the entry data we should preferably do that with `self._abort_if_unique_id_configured` and pass the `updates` parameter.

Reply:
At least in my local testing, it didn't seem to create a new entry if the ID was already configured; that's the reason I just called `async_create_entry` again. It actually seemed to _update_ the original entry. Which I know sounds a little strange, but that's the behavior I observed.

I missed the `updates` parameter to `_abort_if_unique_id_configured`; I didn't look at it that closely since I didn't expect a function with that name would also possibly update an entry. That said, I'm grateful you pointed it out and it would work here.

However, see my comment [here](https://github.com/home-assistant/core/pull/34394#discussion_r443145458) - the unique ID is derived from the user input, and currently the only input we take is the `access_token`. So I'm in a bit of a chicken-and-egg situation unless I can somehow thread the _old_ unique_id from the "broken" config entry into this step in the first place.

Reply:
In latest commits I've refactored how we derive a unique ID for the config entry, so I can actually update the entry appropriately.

I chose to manually call `async_update_entry` and then to abort on my own, since it gave me a little more control over the messaging displayed to the user. Otherwise, calling `_abort_if_unique_id_configured` leads to confusing messaging (and even still, calling `async_abort` even with a custom message is still sub-optimal; we should have a better way to end flows).